test-utils-base/src/commonMain/kotlin/mock/MockMethod.kt

```diff
@@ -0,0 +1,280 @@
+        createStub(matcher).also {
+            stubs[matcher] = it
+        }
+
+    protected fun getStubFor(values: V): S {
+        callParameters.add(values)
+        // First find all the stubs whose matchers match the values received and sort their matchers per parameter in order of strongest constraint.
+        val matchingStubs = stubs.keys.mapNotNull { matchers ->
+            val stub = stubs[matchers]
+            if (ParametersSpec.run { matchers.matches(values) } && stub != null) matchers.asList().sorted() to stub else null
+        }
+        // Ensure there is at least one stub. Otherwise fail
+        if (matchingStubs.isEmpty()) { fail { "No matching stubs found for $values" } }
+        val matchedStubs = (0..matchingStubs.first().first.size).fold(matchingStubs) { remainingMatches, index ->
+            // Find the best matching stub.
+            // Iterate over the length of the number of parameters passed to the method
+            remainingMatches.fold(emptyList()) inner@{ acc, possibleBestMatch ->
+                // If we dont have a best possible match yet, just use the first element.
+                if (acc.isEmpty()) {
+                    return@inner listOf(possibleBestMatch)
+                }
+                // Check whether possibleBestMatch has better, equal, or worse parameter matching for the parameter at index
+                val matcherAtIndexForAcc = acc.first().first[index]
+                val matcherAtIndexForPossibleBestMatch = possibleBestMatch.first[index]
+                when {
+                    matcherAtIndexForAcc < matcherAtIndexForPossibleBestMatch -> acc // Worse, so return acc
+                    matcherAtIndexForAcc == matcherAtIndexForPossibleBestMatch -> acc.toMutableList() + possibleBestMatch // Equal so return acc + possibleBestMatch
+                    else -> listOf(possibleBestMatch) // Better so return possibleBestMatch
+                }
+            }
+        }
+        // Return the first element of the remaining list of stubbs
+        if (matchedStubs.isEmpty()) { fail { "No matching stubs found for $values" } }
+        return matchedStubs.first().second
+    }
+
+    /**
+     * Resets all calls to this mock method. This means that `verify(times=0)` should succeed
+     */
+    fun resetCalls() {
+        callParameters.clear()
+    }
+
+    /**
+     * Removes all stubbs from this mock method. Note that this also removes any default stubbs that may have been created when declaring the mock.
+     */
+    fun resetStubs() {
+        stubs.clear()
+    }
+
+    /**
+     * Resets both calls and stubbs. Shorthand for [resetCalls] and [resetStubs]
+     */
+    fun reset() {
+        resetCalls()
+        resetStubs()
+    }
+
+    internal fun verify(parameters: C, times: Int = 1) {
+        verify(parameters, VerificationRule.times(times))
+    }
+
+    internal fun verify(parameters: C, verificationRule: VerificationRule) = ParametersSpec.apply {
```

Comment:
I think, probably we can use a builder pattern here for parameters and verification rules. 

Would be nice to have something like this:
```
someMethodMock.verify {
      it wasCalled Once
      with parameter SomeParamether
}

anotherMethodMock.verify {
      wasNotCalled within 3 sec
}
```

Reply:
Leaving this for a future improvement